### PR TITLE
Feature/8429 validity state management

### PR DESCRIPTION
**Pull request by nickguendling**
## Description
Introduces the validity state on health certificates

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8429

### CODE REVIEW

**Review by pascal-brause — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift [hunk 1]
```diff
@@ -410,8 +330,91 @@ class HealthCertificateService {
 
 		subscribeToNotifications()
 		updateGradients()
+		updateValidityStates()
 	}
 
+	#if DEBUG
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift:336]: <3

**Review by naveeddotio — APPROVED** [review]

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift [hunk 2]
```diff
@@ -3,33 +3,38 @@
 //
 
 import Foundation
+import OpenCombine
 import HealthCertificateToolkit
 
-struct HealthCertificate: Codable, Equatable, Comparable {
+final class HealthCertificate: Codable, Equatable, Comparable {
```
[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift:9]: Reference semantics make it much easier to pass around the automatically updating state of the health certificate, so I don't think a struct was a good choice here anymore. Would be more difficult to update e.g. the certificate detail screen if it didn't have a reference to its health certificate that automatically updates if the validity state changes.

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift [hunk 3]
```diff
@@ -66,6 +77,16 @@ struct HealthCertificate: Codable, Equatable, Comparable {
...
+
+	@DidSetPublished var validityState: HealthCertificateValidityState {
+		didSet {
+			if validityState != oldValue {
+				objectDidChange.send(self)
+			}
+		}
+	}
+
+	var objectDidChange = OpenCombine.PassthroughSubject<HealthCertificate, Never>()
```
[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift:89]: Good idea, will adjust it it ne next PR as auto merge was enabled on this one

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift:9]: It makes it easier on one side, but complicated on the other side.
Another solution could be to not have the state of validity here, but derive it from the service.

[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift:9]: Well yes, but as we don't want to have a loading state and need to show cells including the validity information quickly on the fly, we would probably need some kind of caching of the validity state in the service. And some kind of notification mechanism if the validity state changes while the user is on the screen/the screen is still in the stack. And to me this seems like the more elegant solution of caching and notifying in contrast to the service providing this as well.

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift:89]: I actually was able to just make it a let, also changed it in the other places where we use this pattern

**Pull request merged**